index.test.ts

```diff
@@ -41,210 +75,216 @@ function getState({value, selectionStart, selectionEnd}: NativeField) {
-	const field = getField('Abacus');
-	let i = 0;
-	replaceFieldText(field, /a/gi, (match, index, string) => {
-		if (i === 0) {
+for (const type of ['textarea', 'input', 'contenteditable'] as const) {
+	test(`${type}: harness test`, t => {
+		t.equal(getState(getField(type)), '|');
+		t.equal(getState(getField(type, '|')), '|');
+		t.equal(getState(getField(type, 'A|')), 'A|');
+		t.equal(getState(getField(type, '|A')), '|A');
+		t.equal(getState(getField(type, '{A}')), '{A}');
+		t.equal(getState(getField(type, 'A{B}C')), 'A{B}C');
+		t.end();
+	});
+
+	test(`${type}: insert() preserves focused item, if focusable`, t => {
+		t.equal(document.activeElement, document.body);
+		const field = getField(type);
+		insertTextIntoField(field, 'A');
+		t.equal(document.activeElement, document.body);
+		t.end();
+	});
+
+	test(`${type}: insert() inserts text in empty field`, t => {
+		const field = getField(type);
+		insertTextIntoField(field, 'a');
+		t.equal(getState(field), 'a|');
+		t.end();
+	});
+
+	test(`${type}: insert() inserts text in empty input element`, t => {
+		const field = getField(type, '|');
+		insertTextIntoField(field, 'a');
+		t.equal(getState(field), 'a|');
+		t.end();
+	});
+
+	test(`${type}: insert() appends text to unselected field`, t => {
+		const field = getField(type, 'W|');
+		insertTextIntoField(field, 'O');
+		t.equal(getState(field), 'WO|');
+		t.end();
+	});
+
+	test(`${type}: insert() inserts text in the middle`, t => {
+		const field = getField(type, 'W|O');
+		insertTextIntoField(field, 'A');
+		t.equal(getState(field), 'WA|O');
+		t.end();
+	});
+
+	test(`${type}: insert() replaces selected text`, t => {
+		const field = getField(type, '{W}O');
+		insertTextIntoField(field, 'A');
+		t.equal(getState(field), 'A|O');
+		t.end();
+	});
+
+	test(`${type}: insert() replaces selected text even if string is ""`, t => {
+		const field = getField(type, '{W}O');
+		insertTextIntoField(field, '');
+		t.equal(getState(field), '|O');
+		t.end();
+	});
+
+	test(`${type}: insert() fires input event`, t => {
+		const field = getField(type);
+		field.addEventListener('input', event => {
+			t.equal(event.type, 'input');
+			// TODO: t.equal(event.inputType, 'insert');
+			t.end();
+		});
+		insertTextIntoField(field, 'A');
+	});
+
+	test(`${type}: set() replaces the whole content`, t => {
+		const field = getField(type, '{W}O');
+		setFieldText(field, 'ABC');
+		t.equal(getState(field), 'ABC|');
+		t.end();
+	});
+
+	test(`${type}: getSelection()`, t => {
+		const field = getField(type, 'W{O}A');
+		t.equal(getFieldSelection(field), 'O');
+		t.end();
+	});
+
+	test(`${type}: getSelection() without selection`, t => {
+		const field = getField(type, 'WOA|');
+		t.equal(getFieldSelection(field), '');
+		t.end();
+	});
+
+	test(`${type}: wrapSelection() wraps selected text`, t => {
+		const field = getField(type, 'W{O}A');
+		wrapFieldSelection(field, '*');
+		t.equal(getState(field), 'W*{O}*A');
```

Comment:
Kinda tough here. The test is failing because the selection isn't being restored correctly. I think I'll have to move some range helpers from https://github.com/fregante/zip-text-nodes/blob/a73cd33b5e328be4a2afe5ee406da93606c95c55/index.ts#L1-L46

Basically I'd need to extract `selectionStart`/`selectionEnd` from a `contentEditable`, and set it again the same exact way I'm doing for the regular inputs:

https://github.com/bfred-it/insert-text-textarea/blob/9670bd31d0f1b084e55c6aa878b98c665469c70f/index.ts/#L95-L97